src/codeGenerator/writer.ts

```diff
@@ -0,0 +1,70 @@
+const fs = require("fs");
+const path = require("path");
+
+import { AdvancementValue, FunctionValue } from "./store";
+
+export abstract class Serializable {
+  public abstract serialize(): any;
```

Comment:
Is there a better type to return here? AFAIK this always returns an `object` of some kind, right?

Reply:
Yes this would expect to return any `object`. If we were to be more specific in terms of return type here, it seems out of place as it would mean that `Serializable` would need to "know" about implementation details of how the caller wishes to return things?

We could be more specific in defining return types where implementers use `serialize()` but:
1. I'm too lazy to define every possible parameter that an Advancement  or a Function could take (with dependencies on existence of keys)
2. If we are only defining the subset of parameters that we currently use, this could readily change in the future so we should do it all (if so, refer to 1) or not at all (and leave the return type as `any`)

Reply:
Hmm, fair enough. On the point about "know"ing about implementation details; I don't think we need to worry about that. This is because we know it will always return an `object`, so would it make sense to return `object` instead of `any`? Is there a case where `Serializable` might not return `object`?

If so, I'm fine with this

Reply:
yeah, for mc functions, the final output is just a new-line separated string (representing commands), whereas advancements are JSON (both of which implement `Serializable`)